src/mapgen/mg_biome.cpp

```diff
@@ -135,13 +124,15 @@ BiomeGenOriginal::BiomeGenOriginal(BiomeManager *biomemgr,
 	// is disabled.
 	memset(biomemap, 0, sizeof(biome_t) * m_csize.X * m_csize.Z);
 
-	// Calculate cache of Y transition points
+	// Calculate cache of Y transition points. For downwards scanning only!
 	std::vector<s16> values;
 	values.reserve(m_bmgr->getNumObjects() * 2);
 	for (size_t i = 0; i < m_bmgr->getNumObjects(); i++) {
 		Biome *b = (Biome *)m_bmgr->getRaw(i);
 		values.push_back(b->max_pos.Y);
-		values.push_back(b->min_pos.Y);
+		// Optimization: biomes often have perfect Y transitions, resulting in
+		// unnecessary Y pairs. i.e. when (biome A min Y) == (biome B max Y + 1)
+		values.push_back(std::max<s32>((s32)b->min_pos.Y - 1, -MAX_MAP_GENERATION_LIMIT));
```

Comment:
we have ``std::unique`` for that already?

Reply:
`max_pos.Y` and `min_pos.Y` are both inclusive, which means that in the following example with 2 biomes (A (above) and B (below)), there are transitions which we don't care about
```
y = max_pos.Y  (first position of biome A) IMPORTANT
 X
 X
 (span of A)
 X
 X
y = min_pos.Y  (last position of biome A) (superfluous)
y = max_pos.Y  (first position of biome B) IMPORTANT
 X
 X
 (span of B)
 X
 X
y = min_pos.Y  (last position of biome B) (superfluous)
```

Reply:
as a practical example we collect positions
`{3000, 2000, 2000, 1000, 1000, 0, 0, -31000}`
and then by running unique on it we get
`{3000, 2000, 1000, 0, -31000}`
or do you mean something else?

Reply:
That is correct. Before you might have
```
{3000, 256, 255, 1, 0, -255, -256, -31000 }
 ^       ^  ^    ^  ^     ^  ^          ^
 ^^^^^^^^^  |    |  |     |  |          |  biome A
            ^^^^^^  |     |  |          |  biome B
                    ^^^^^^^  |          |  biome C
                             ^^^^^^^^^^^^  biome D
```
which can be reduced to
```
{3000, 255, 0, -256 }
 ^     ^    ^  ^     
 ^^^^^^|    |  |     biome A
       ^^^^^|  |     biome B
            ^^^|     biome C
               ^^^^^ biome D (until < -31000)
```
Since we're generating biomes from top to bottom, we only need to fetch the new update at the uppermost coordinate (or also lowermost coordinate if the biomes don't stack perfectly and there's a gap).